pallets/stateful-storage/src/types.rs

```diff
@@ -0,0 +1,273 @@
+	/// Message Source Account identifier
+	#[codec(compact)]
+	pub msa_id: MessageSourceId,
+
+	/// Schema id of this storage
+	#[codec(compact)]
+	pub schema_id: SchemaId,
+
+	/// Hash of targeted page to avoid race conditions
+	#[codec(compact)]
+	pub target_hash: PageHash,
+
+	/// The block number at which the signed proof will expire
+	pub expiration: T::BlockNumber,
+
+	/// actions to apply to storage
+	pub actions: BoundedVec<ItemAction, <T as Config>::MaxItemizedActionsCount>,
+}
+
+/// Payload containing all necessary fields to verify signatures to upsert a Paginated storage
+#[derive(Encode, Decode, TypeInfo, MaxEncodedLen, PartialEq, RuntimeDebugNoBound, Clone)]
+#[scale_info(skip_type_params(T))]
+pub struct PaginatedUpsertSignaturePayload<T: Config> {
+	/// Message Source Account identifier
+	#[codec(compact)]
+	pub msa_id: MessageSourceId,
+
+	/// Schema id of this storage
+	#[codec(compact)]
+	pub schema_id: SchemaId,
+
+	/// Page id of this storage
+	#[codec(compact)]
+	pub page_id: PageId,
+
+	/// Hash of targeted page to avoid race conditions
+	#[codec(compact)]
+	pub target_hash: PageHash,
+
+	/// The block number at which the signed proof will expire
+	pub expiration: T::BlockNumber,
+
+	/// payload to update the page with
+	pub payload: BoundedVec<u8, <T as Config>::MaxPaginatedPageSizeBytes>,
+}
+
+/// Payload containing all necessary fields to verify signatures to delete a Paginated storage
+#[derive(Encode, Decode, TypeInfo, MaxEncodedLen, PartialEq, RuntimeDebugNoBound, Clone)]
+#[scale_info(skip_type_params(T))]
+pub struct PaginatedDeleteSignaturePayload<T: Config> {
+	/// Message Source Account identifier
+	#[codec(compact)]
+	pub msa_id: MessageSourceId,
+
+	/// Schema id of this storage
+	#[codec(compact)]
+	pub schema_id: SchemaId,
+
+	/// Page id of this storage
+	#[codec(compact)]
+	pub page_id: PageId,
+
+	/// Hash of targeted page to avoid race conditions
+	#[codec(compact)]
+	pub target_hash: PageHash,
+
+	/// The block number at which the signed proof will expire
+	pub expiration: T::BlockNumber,
+}
+
+/// A generic page of data which supports both Itemized and Paginated
+#[derive(Encode, Decode, TypeInfo, MaxEncodedLen, Debug, Default)]
+#[scale_info(skip_type_params(PageDataSize))]
+#[codec(mel_bound(PageDataSize: MaxEncodedLen))]
+pub struct Page<PageDataSize: Get<u32>> {
+	pub data: BoundedVec<u8, PageDataSize>,
+}
+
+/// An internal struct which contains the parsed items in a page
+#[derive(Debug, PartialEq)]
+pub struct ParsedItemPage<'a> {
+	/// page current size
+	pub page_size: usize,
+	/// a map of item index to a slice of blob (including header is optional)
+	pub items: BTreeMap<u16, &'a [u8]>,
+}
+
+impl<PageDataSize: Get<u32>> Page<PageDataSize> {
+	pub fn is_empty(&self) -> bool {
+		self.data.is_empty()
+	}
+
+	pub fn get_hash(&self) -> PageHash {
+		if self.is_empty() {
+			return PageHash::default()
+		}
+		let mut hasher = XxHash32::with_seed(0);
```

Comment:
Is it worth switching to XXH3 and then just taking the first 32 bits of it? It is a lot faster. https://crates.io/crates/twox-hash and https://cyan4973.github.io/xxHash/

Reply:
Modified to use xxh3 first 4 bytes